### PR TITLE
#205. Eliminate sleeping locks in Tempesta sockets API code

### PR DESCRIPTION
**Pull request by avbelov23**

### CODE REVIEW

File: tempesta_fw/sock.c [hunk 1]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+	int err;
+
+	old_state = sk->sk_state;
+	if (!((1 << old_state) & (TCPF_CLOSE | TCPF_LISTEN)))
+		goto out;
+
+	/* Really, if the socket is already in listen state
+	 * we can only allow the backlog to be adjusted.
+	 */
+	if (old_state != TCP_LISTEN) {
```
Comment (keshonok) [inline tempesta_fw/sock.c]: Do we not want to support TCP Fast Open?
If we really don't, then we should at least have a comment about that here.

File: tempesta_fw/sock.c [hunk 2]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
-		.type = sk->sk_type,
-		.state = SS_UNCONNECTED
-	};
+	int err;
+
 	BUG_ON(sk->sk_type != SOCK_STREAM);
-	return inet_listen(&sock, backlog);
+
+	bh_lock_sock(sk);
+	err = ss_inet_listen(sk, backlog);
```
Comment (keshonok) [inline tempesta_fw/sock.c]: `ss_inet_listen()` is never called in any context other than from `ss_listen()`. Therefore there's no reason to have it as a separate function

File: tempesta_fw/sock.c [hunk 3]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
-	release_sock(sk);
+	bh_unlock_sock(sk);
 	return err;
 }
 EXPORT_SYMBOL(ss_connect);
 
 /*
- * The original functions are inet_bind() and inet6_bind().
- * These two can be made a bit shorter should that become necessary.
+ * Bind for IPv4 protocol. The original functions are inet_bind().
```
Comment (keshonok) [inline tempesta_fw/sock.c]: The original function~~s~~ ~~are~~ is inet_bind().

File: tempesta_fw/sock.c [hunk 4]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+ *
+ * NOTE: This code assumes that both IPv4 and IPv6 are compiled in as
+ * part of the Linux kernel, and not as separate loadable kernel modules.
```
Comment (keshonok) [inline tempesta_fw/sock.c]: I don't think this comment is relevant here (as well as below). It's relevant for `ss_inet_create()` as the original implementation tries to load the module if the code is not in the kernel yet, while the Sync Sockets implementation skips that.

File: tempesta_fw/sock.c [hunk 5]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+	inet->inet_daddr = 0;
+	inet->inet_dport = 0;
+	sk_dst_reset(sk);
+	err = 0;
+out:
+	return err;
+}
+
+/*
+ * Bind for IPv6 protocol. The original functions inet6_bind().
```
Comment (keshonok) [inline tempesta_fw/sock.c]: The original function~~S~~ **is** inet6_bind().

File: tempesta_fw/sock.c [hunk 6]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+ *
+ * NOTE: This code assumes that both IPv4 and IPv6 are compiled in as
+ * part of the Linux kernel, and not as separate loadable kernel modules.
```
Comment (keshonok) [inline tempesta_fw/sock.c]: Please the comment above on the same issue.

File: tempesta_fw/sock.c [hunk 7]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+	 *  is temporarily down)
+	 */
+	err = -EADDRNOTAVAIL;
+	if (!sysctl_ip_nonlocal_bind &&
+	    !(inet->freebind || inet->transparent) &&
+	    addr->sin_addr.s_addr != htonl(INADDR_ANY) &&
+	    chk_addr_ret != RTN_LOCAL &&
+	    chk_addr_ret != RTN_MULTICAST &&
+	    chk_addr_ret != RTN_BROADCAST)
+		goto out;
```
Comment (keshonok) [inline tempesta_fw/sock.c]: As a matter of fact, we do set `inet_sk(sk)->freebind = 1;` in `sock_clnt.c`, so the condition above is never true. Do we need to keep this code block?

File: tempesta_fw/sock.c [hunk 8]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+	    !ns_capable(net->user_ns, CAP_NET_BIND_SERVICE))
+		goto out;
+
+	/*      We keep a pair of addresses. rcv_saddr is the one
+	 *      used by hash lookups, and saddr is used for transmit.
+	 *
+	 *      In the BSD API these are the same except where it
+	 *      would be illegal to use them (multicast/broadcast) in
+	 *      which case the sending device address is used.
+	 */
```
Comment (keshonok) [inline tempesta_fw/sock.c]: I believe that there's no need to copy this kind of elaborate comments from original function here. We have a reference to the original function, and that should be sufficient. There are big blocks of comments in other places as well.

File: tempesta_fw/sock.c [hunk 9]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+	    addr->sin_addr.s_addr != htonl(INADDR_ANY) &&
+	    chk_addr_ret != RTN_LOCAL &&
+	    chk_addr_ret != RTN_MULTICAST &&
+	    chk_addr_ret != RTN_BROADCAST)
+		goto out;
+
+	snum = ntohs(addr->sin_port);
+	err = -EACCES;
+	if (snum && snum < PROT_SOCK &&
+	    !ns_capable(net->user_ns, CAP_NET_BIND_SERVICE))
```
Comment (keshonok) [inline tempesta_fw/sock.c]: This check surely doesn't hurt, but do we need it?

File: tempesta_fw/sock.c [hunk 10]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+	 *      which case the sending device address is used.
+	 */
+
+	/* Check these errors (active socket, double bind). */
+	err = -EINVAL;
+	if (sk->sk_state != TCP_CLOSE || inet->inet_num)
+		goto out;
+
+	inet->inet_rcv_saddr = inet->inet_saddr = addr->sin_addr.s_addr;
+	if (chk_addr_ret == RTN_MULTICAST || chk_addr_ret == RTN_BROADCAST)
```
Comment (keshonok) [inline tempesta_fw/sock.c]: Do we need this check?

File: tempesta_fw/sock.c [hunk 11]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+	if (snum)
+		sk->sk_userlocks |= SOCK_BINDPORT_LOCK;
+	inet->inet_sport = htons(inet->inet_num);
+	inet->inet_dport = 0;
+	inet->inet_daddr = 0;
+unlock:
+	rcu_read_unlock();
+out:
+	return err;
+}
```
Comment (keshonok) [inline tempesta_fw/sock.c]: Basically, most questions here are similar to questions to `ss_inet_bind()` above.

File: tempesta_fw/sock.c [hunk 12]
```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
...
+
+	if (addr_len < SIN6_LEN_RFC2133)
+		return -EINVAL;
+
+	addr_type = ipv6_addr_type(&addr->sin6_addr);
+	if ((addr_type & IPV6_ADDR_MULTICAST))
+		return -EINVAL;
+
+	snum = ntohs(addr->sin6_port);
+	if (snum && snum < PROT_SOCK && !ns_capable(net->user_ns, CAP_NET_BIND_SERVICE))
```
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: Too long line, please keep 80 characters as maximum line length.

**Pull request merged**